graph/api/src/main/scala/org/apache/spark/graph/api/GraphElementFrame.scala

```diff
@@ -0,0 +1,260 @@
+ */
+abstract class GraphElementFrame {
+
+  /**
+   * Initial DataFrame that can still contain unmapped, arbitrarily ordered columns.
+   *
+   * @since 3.0.0
+   */
+  def df: DataFrame
+
+  /**
+   * Name of the column that contains the graph element identifier.
+   *
+   * @since 3.0.0
+   */
+  def idColumn: String
+
+  /**
+   * Name of all columns that contain graph element identifiers.
+   *
+   * @since 3.0.0
+   */
+  def idColumns: Seq[String] = Seq(idColumn)
+
+  /**
+   * Mapping from graph element property keys to the columns that contain the corresponding property
+   * values.
+   *
+   * @since 3.0.0
+   */
+  def properties: Map[String, String]
+
+}
+
+object NodeFrame {
+
+  /**
+   * Describes how to map an initial DataFrame to nodes.
+   *
+   * All columns apart from the given `idColumn` are mapped to node properties.
+   *
+   * @param df        DataFrame containing a single node in each row
+   * @param idColumn  column that contains the node identifier
+   * @param labelSet  labels that are assigned to all nodes
+   * @since 3.0.0
+   */
+  def create(df: DataFrame, idColumn: String, labelSet: Set[String]): NodeFrame = {
+    val properties = (df.columns.toSet - idColumn)
+      .map(columnName => columnName -> columnName)
+      .toMap
+    create(df, idColumn, labelSet, properties)
+  }
+
+  /**
+   * Describes how to map an initial DataFrame to nodes.
+   *
+   * All columns apart from the given `idColumn` are mapped to node properties.
+   *
+   * @param df        DataFrame containing a single node in each row
+   * @param idColumn  column that contains the node identifier
+   * @param labelSet  labels that are assigned to all nodes
+   * @param properties mapping from property keys to corresponding columns
+   * @since 3.0.0
+   */
+  def create(
+      df: DataFrame,
+      idColumn: String,
+      labelSet: Set[String],
+      properties: Map[String, String]): NodeFrame = {
+    NodeFrame(df, idColumn, labelSet, properties)
+  }
+
+  /**
+   * Describes how to map an initial DataFrame to nodes.
+   *
+   * All columns apart from the given `idColumn` are mapped to node properties.
+   *
+   * @param df        DataFrame containing a single node in each row
+   * @param idColumn  column that contains the node identifier
+   * @param labelSet  labels that are assigned to all nodes
+   * @since 3.0.0
+   */
+  def create(df: DataFrame, idColumn: String, labelSet: java.util.Set[String]): NodeFrame = {
+    create(df, idColumn, labelSet.asScala.toSet)
+  }
+
+  /**
+   * Describes how to map an initial DataFrame to nodes.
+   *
+   * All columns apart from the given `idColumn` are mapped to node properties.
+   *
+   * @param df        DataFrame containing a single node in each row
+   * @param idColumn  column that contains the node identifier
+   * @param labelSet  labels that are assigned to all nodes
+   * @param properties mapping from property keys to corresponding columns
+   * @since 3.0.0
+   */
+  def create(
```

Comment:
Should consider builder pattern to avoid many overloading methods. For example:

~~~scala
sess.buildNodeFrame
  .dataset(...)
  .id(..)
  .labels(..) // <- use Array to support both Scala and Java
  .props(...) // <- overload this one for Java
  .build()
~~~

Reply:
We actually discussed that at one point as we do this in Morpheus: https://github.com/opencypher/morpheus/blob/master/okapi-api/src/main/scala/org/opencypher/okapi/api/io/conversion/NodeMappingBuilder.scala

We decided to not do it, in order to minimize the number of classes we need to map in Python.

Do you think we should add it?

Reply:
Yep. Please follow @mengxr 's advice.

Reply:
Done in 0736618 and 779e1ff